calabash/Classes/FranklyServer/Routes/LPClearTextRoute.m

```diff
@@ -0,0 +1,140 @@
+#if ! __has_feature(objc_arc)
+#warning This file must be compiled with ARC. Use -fobjc-arc flag (or convert project to ARC).
+#endif
+
+#import "LPClearTextRoute.h"
+#import "UIScriptParser.h"
+#import "LPTouchUtils.h"
+#import "LPCocoaLumberjack.h"
+#import "LPJSONUtils.h"
+
+@implementation LPClearTextRoute
+
+- (BOOL) supportsMethod:(NSString *) method atPath:(NSString *) path {
+  return [method isEqualToString:@"GET"];
+}
+
+- (NSDictionary *)failureResponseWithReason:(NSString *)reason {
+    return @{
+        @"outcome" : @"FAILURE",
+        @"reason" : reason,
+        @"details" : @""
+    };
+}
+
+- (NSDictionary *)successResponseWithResult:(id)result {
+    return @{
+       @"outcome" : @"SUCCESS",
+       @"results" : @[result]
+    };
+}
+
+- (id)firstResponder {
+  NSString *query = @"* isFirstResponder:1 index:0";
+  UIScriptParser *parser = [[UIScriptParser alloc] initWithUIScript:query];
+  [parser parse];
+  NSArray *allWindows = [LPTouchUtils applicationWindows];
+  NSArray *results = [parser evalWith:allWindows];
+  if (!results || results.count == 0) {
+    return nil;
+  } else {
+    return results[0];
+  }
+}
+
+- (NSDictionary *)JSONResponseForMethod:(NSString *) method
+                                     URI:(NSString *) path
+                                    data:(NSDictionary *) data {
+  NSString *reason = @"";
+  id target = [self firstResponder];
+  if (!target) {
+    reason = [NSString stringWithFormat:@"Cannot clear text because no view is first responder"];
+    LPLogError(@"%@", reason);
+    return [self failureResponseWithReason:reason];
+  }
+
+  if (![target respondsToSelector:@selector(isFirstResponder)]) {
+    reason = [NSString stringWithFormat:@"Target %@ does not respond to 'isFirstResponder'", target];
+    return [self failureResponseWithReason:reason];
+  }
+
+  id delegate = nil;
+  if ([target respondsToSelector:@selector(delegate)]) {
+    delegate = [target performSelector:@selector(delegate)];
+  }
+
+  BOOL shouldClearText = YES;
+  NSString *existingText = @"";
+  if ([target respondsToSelector:@selector(text)]) {
+    id currentText = [target performSelector:@selector(text)];
+    if (currentText) {
+      existingText = (NSString *)currentText;
+    }
+  }
+
+  NSRange textRange = NSRangeFromString(existingText);
+  if ([target respondsToSelector:@selector(setText:)]) {
+    shouldClearText = [self shouldClearTextInRange:textRange WithDelegate:delegate withTarget:target];
+    if (shouldClearText) {
+      [target performSelector:@selector(setText:) withObject:@""];
+    }
+  } else {
+    // Check if it's UIKeyInput protocol
+    LPLogDebug(@"Target class %@ does not respond to setText checking UIKeyInput", target);
+    if ([target conformsToProtocol:@protocol(UIKeyInput)]) {
+      LPLogDebug(@"Attempting to clearText with deleteBackward");
+      while ([target hasText]) {
```

Comment:
You definitely need to pause after each `[target deleteBackward]` call.

```
# 0.1 is a suggestion - I think it makes sense.
[target deleteBackward];
CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.1, false);
```

Ah, right - UIKeyInput does not define a way to ask for text.

We have two options: time out or cap the number of characters.

A cap of 1000 characters is probably sufficient.

A timeout of 20 seconds is probably sufficient.

If you opt for timeout, please use:

* https://github.com/calabash/DeviceAgent.iOS/blob/develop/Server/Utilities/CBXMachClock.h

You can copy the file over and change the prefix to `LP`.

Follow this pattern for timing out:

```
    NSTimeInterval startTime = [[CBXMachClock sharedClock] absoluteTime];
    NSTimeInterval endTime = startTime + 20.0;

     while ([target hasText] && [[CBXMachClock sharedClock] absoluteTime] < endTime) {
         [self deleteBackwards];
         CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0.1, false);
     }
 
   NSTimeInterval elapsed = [[CBXMachClock sharedClock] absoluteTime] - startTime;
   if ([target hasText) {
     // Log and fail
   } else {
    // Log and pass
  }
```